src/ophyd_async/core/device.py

```diff
@@ -83,12 +93,18 @@ async def connect(self, mock: bool = False, timeout: float = DEFAULT_TIMEOUT):
         timeout:
             Time to wait before failing with a TimeoutError.
         """
-        coros = {
-            name: child_device.connect(mock=mock, timeout=timeout)
-            for name, child_device in self.children()
-        }
-        if coros:
-            await wait_for_connection(**coros)
+
+        if force_reconnect or not self._previous_connect_success:
```

Comment:
Unfortunately this misses a requirement of the original ticket, if we `connect()` and there is already another `connect()` running then we should `await connect_task` from the previous connect.

@evalott100 we should also think about what happens when switching between `connect(mock=True)` and `connect(mock=False)`. Should we ban it, or do a `force_reconnect` each time we change the argument `mock`?

Reply:
In my opinion it would be quite useful during testing to be able to make a signal fail to connect and then swap in `mock=True` on the next connect. In the other rarer case of a test using an ioc/softioc, I don't think it's neccessary to prohibit reconnection with mock.